model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/CacheManager.java

```diff
@@ -132,6 +137,55 @@ public Object invalidateObject(String id) {
         return removed;
     }
 
+    /**
+     * This method is intended to allow for the invalidation of entire sets of keys within the cache that match
+     * the given prefix. Rather than iterating through each object in the cache and dropping based on ID lookups,
+     * find all the keys that match a composable prefix and drop everything after it.
+     * @param prefix The prefix that should be used for matching
+     * @return A set of objects representing all removed items from the cache
+     */
+    public Set<Object> invalidatePrefix(String prefix) {
+        Set<String> keys = cache.keySet().parallelStream().filter(key -> key.startsWith(prefix)).collect(Collectors.toSet());
```

Comment:
This way of working with the cache entries is specifically recommended by infinispan https://infinispan.org/docs/stable/titles/embedding/embedding.html#streams 

The `parallelStream()` method is overloaded to return their own implementation that works well with the cache store

Reply:
Are you sure this is OK? AFAIK this can severely limit processing of concurrent requests when the thread pool for parallel streams is exhausted.

Reply:
That sounds fine to me! Agree that it would make debugging harder. I know keycloak technically supports having multiple distributed nodes for the cache (though I don't know how widely that's used) which was why I reached for parallel stream... though looking at the docs again I think I have partially misunderstood how that system works. I'll change over to a normal stream!